crates/driver/src/domain/competition/score.rs

```diff
@@ -76,31 +89,102 @@ impl From<eth::U256> for ObjectiveValue {
     }
 }
 
-/// Comparing scores and objective values is needed to make sure the score is
-/// not higher than the objective value, which is a requirement for the score to
-/// be valid.
-impl std::cmp::PartialEq<ObjectiveValue> for Score {
-    fn eq(&self, other: &ObjectiveValue) -> bool {
-        self.0.eq(&other.0)
+impl std::ops::Add for ObjectiveValue {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        Self(self.0 + other.0)
     }
 }
 
-impl std::cmp::PartialOrd<ObjectiveValue> for Score {
-    fn partial_cmp(&self, other: &ObjectiveValue) -> Option<Ordering> {
-        self.0.partial_cmp(&other.0)
+impl num::Zero for ObjectiveValue {
+    fn zero() -> Self {
+        Self(eth::U256::zero())
+    }
+
+    fn is_zero(&self) -> bool {
+        self.0.is_zero()
     }
 }
 
 #[derive(Debug, thiserror::Error)]
 pub enum Error {
-    #[error("objective value is non-positive")]
-    ObjectiveValueNonPositive,
+    /// The solution has zero score. Zero score solutions are not allowed as per
+    /// CIP20 definition. The main reason being that reference score is zero,
+    /// and if only one solution is in competition with zero score, that
+    /// solution would receive 0 reward (reward = score - reference score).
     #[error("score is zero")]
     ZeroScore,
-    #[error("objective value {0:?} is higher than the objective {1:?}")]
-    ScoreHigherThanObjective(Score, ObjectiveValue),
-    #[error("success probability is out of range {0:?}")]
-    SuccessProbabilityOutOfRange(SuccessProbability),
-    #[error("invalid objective value")]
+    /// Protocol does not allow solutions that are claimed to be "better" than
+    /// the actual value they bring (quality). It is expected that score
+    /// is always lower than quality, because there is always some
+    /// execution cost that needs to be incorporated into the score and lower
+    /// it.
+    #[error("score {0:?} is higher than the quality {1:?}")]
+    ScoreHigherThanQuality(Score, Quality),
+    /// Errors only applicable to scores that use success probability.
+    #[error(transparent)]
+    RiskAdjusted(#[from] risk::Error),
+    #[error(transparent)]
     Boundary(#[from] boundary::Error),
 }
+
+pub mod risk {
+    /// Contains functionality and error types for scores that are based on
+    /// success probability.
```

Comment:
nit: Should probably be a doc comment.
```suggestion
    //! Contains functionality and error types for scores that are based on
    //! success probability.
```